addons/hr_homeworking/models/hr_homeworking.py

```diff
@@ -0,0 +1,306 @@
+            'work_location_id': work,
+            'employee_id': employee_id,
+            'start_date': date,
+        })
+        return True
+
+    def exceptional_worklocation_to_default_worklocation(self, vals, unlink=True):
+        default_date = vals.get("start_date")
+        if default_date:
+            default_date = string_to_datetime(default_date).date()
+        else:
+            default_date = self.start_date
+        date = default_date
+        default_work = self.env['hr.employee.location'].search([
+            ("employee_id.id", "=", self.employee_id.id),
+            ("weekday", "=", default_date.weekday()),
+            ("weekly", "=", True),
+            ("current", "=", True)
+        ])
+        removed_work = self.env['hr.employee.location'].search([
+            ("employee_id.id", "=", self.employee_id.id),
+            ("start_date", "=", date),
+            ("removed", "=", True),
+        ])
+        if removed_work:
+            removed_work.unlink()
+        if default_date < fields.Date().today():
+            default_date = self._next_weekday_date_after_today(date)
+        if default_work:
+            default_work._save_previous_default_worklocation()
+        if date != default_date:
+            self.env['hr.employee.location'].create({
+                'employee_id': self.employee_id.id,
+                'start_date': default_date,
+                'work_location_id': vals.get("work_location_id", self.work_location_id.id),
+                'weekly': True
+            })
+            vals["weekly"] = False
+        if unlink:
+            self.unlink()
+        else:
+            if vals.get("work_location_id"):
+                self.work_location_id = vals["work_location_id"]
+        return vals
+
+    def add_removed_work_location(self, date):
+        self.env['hr.employee.location'].create({
+            'work_location_id': self.work_location_id.id,
+            'employee_id': self.employee_id.id,
+            'start_date': date,
+            'removed': True,
+            'parent_default_homeworking': self.id,
+        })
+
+    def delete_default_worklocation(self):
+        self.ensure_one()
+        self._save_previous_default_worklocation()
+
+    def _get_new_worklocation_vals(self, vals):
+        date = vals.get("start_date")
+        except_vals = {
+            "start_date": date,
+            "work_location_id": vals.get("work_location_id", self.work_location_id.id),
+            "weekly": vals.get("weekly", self.weekly),
+            "employee_id": vals.get("employee_id", self.employee_id.id),
+        }
+        # to not rewrite the past
+        self._save_previous_default_worklocation()
+        values = vals
+        if date < fields.Date().today():
+            self.env['hr.employee.location'].add_exceptional_worklocation(except_vals)
+            values["start_date"] = self.env['hr.employee.location']._next_weekday_date_after_today(date)
```

Comment:
This will still modify the value in `vals`
You need to do `values = vals.copy()` to have another copy of the original dict